core/state_processor.go

```diff
@@ -202,3 +206,38 @@ func ProcessBeaconBlockRoot(beaconRoot common.Hash, vmenv *vm.EVM, statedb *stat
 	_, _, _ = vmenv.Call(vm.AccountRef(msg.From), *msg.To, msg.Data, 30_000_000, common.U2560)
 	statedb.Finalise(true)
 }
+
+// ProcessBlockHashHistory is called at every block to insert the parent block hash
+// in the history storage contract as per EIP-2935. At the EIP-2935 fork block, it
+// populates the whole buffer with block hashes.
+func ProcessBlockHashHistory(statedb *state.StateDB, header *types.Header, chainConfig *params.ChainConfig, chain consensus.ChainHeaderReader) {
+	var (
+		prevHash   = header.ParentHash
+		parent     = chain.GetHeaderByHash(prevHash)
+		prevNumber = parent.Number.Uint64()
+	)
+	ProcessParentBlockHash(statedb, prevNumber, prevHash)
+	// History already inserted.
+	if chainConfig.IsPrague(parent.Number, parent.Time) || prevNumber == 0 {
+		return
+	}
+	var low uint64
+	if prevNumber > 255 {
+		low = prevNumber - 255
+	}
+	for i := prevNumber - 1; i >= low; i-- {
+		ProcessParentBlockHash(statedb, i, parent.ParentHash)
+		parent = chain.GetHeader(parent.ParentHash, i)
+	}
+}
+
+// ProcessParentBlockHash stores the parent block hash in the history storage contract
+// as per EIP-2935.
+func ProcessParentBlockHash(statedb *state.StateDB, prevNumber uint64, prevHash common.Hash) {
+	ringIndex := prevNumber % 256
+	var key common.Hash
+	binary.BigEndian.PutUint64(key[24:], ringIndex)
+	statedb.SetState(params.HistoryStorageAddress, key, prevHash)
+	statedb.AddAddressToAccessList(params.HistoryStorageAddress)
+	statedb.AddSlotToAccessList(params.HistoryStorageAddress, key)
```

Comment:
Why do you do this? I don't thin this is correct. At best, it's moot. 

We do this processing before any transactions. The access list is per-transaction. 

In the worst case, you are now modifying the access list for the first transaction. 
In the best case, this access-list change is reset once we start executing transcations. 
But maybe you indeed intended for the address/slot to be 'hot' for all transactions? If so, this is not the way to do it.

Reply:
Yes you're right. When I transferred this code from the verkle branch where they have to touch the accounts/slots for the block witness I didn't think it's different from mainnet. Will fix.

Reply:
Semi-related: I know this is currently the spec, but I don't think we should be reaching into the state like this. We have the system call pattern from 4788 already, I think we should update the spec to mirror it.

Reply:
Fixed. BTW also the 4788 processing added the address to ACL. Removed that too.